src/parser/shared/normalizers/Channeling.ts

```diff
@@ -0,0 +1,344 @@
+    buffChannelSpec(SPELLS.ESSENCE_FONT.id),
+    buffChannelSpec(SPELLS.SOOTHING_MIST.id),
+    buffChannelSpec(SPELLS.CRACKLING_JADE_LIGHTNING.id),
+    buffChannelSpec(SPELLS.FISTS_OF_FURY_CAST.id),
+    // Demon Hunter
+    buffChannelSpec(SPELLS.EYE_BEAM.id), // TODO special handling because of the two buffs?
+    // Shaman
+    // Hunter
+    buffChannelSpec(SPELLS.RAPID_FIRE.id),
+    // TODO Barrage - can't find a single log with someone taking the talent...
+    // Paladin
+    // Warrior
+    buffChannelSpec(SPELLS.BLADESTORM_TALENT.id),
+    buffChannelSpec(SPELLS.BLADESTORM.id),
+    // Death Knight
+    buffChannelSpec(SPELLS.BLOODDRINKER_TALENT.id),
+  ];
+
+  // registered special case handlers, mapped by guid
+  channelSpecMap: { [key: number]: ChannelHandler } = {};
+
+  /**
+   * Constructs a Channeling normalizer that deals with normal hardcasts and instants by default,
+   * and then allows special case functions to be registered by spellId for channels.
+   * @param channelSpecs special case functions that handle specific spells
+   */
+  constructor(options: Options) {
+    super(options);
+
+    // populate the specs into a mapping for quick handling
+    Channeling.CHANNEL_SPECS.forEach((spec) => {
+      spec.guids.forEach((guid) => {
+        if (this.channelSpecMap[guid]) {
+          console.error(
+            'Channeling module trying to add more than one handler for spell guid ' +
+              guid +
+              ' - only the latest handler will apply!',
+          );
+        }
+        this.channelSpecMap[guid] = spec.handler;
+      });
+    });
+  }
+
+  normalize(events: AnyEvent[]) {
+    const eventsInserter: InsertableEventsWrapper = new InsertableEventsWrapper(events);
+
+    const channelState: ChannelState = {
+      unresolvedChannel: null,
+      eventsInserter,
+    };
+
+    // for each event, check if there is a special case handler for that GUID -
+    // if so - call the handler, if not - call the default handling
+    events.forEach((event: AnyEvent, index: number) => {
+      if (!HasSource(event) || event.sourceID !== this.selectedCombatant.id) {
+        return; // only concerned with events coming from selected player
+      }
+      const handler = HasAbility(event) ? this.channelSpecMap[event.ability.guid] : undefined;
+      if (handler) {
+        handler(event, events, index, channelState);
+      } else {
+        this.defaultHandler(event, events, index, channelState);
+      }
+    });
+
+    return eventsInserter.build();
+  }
+
+  /**
+   * Handles the default cases of regular cast time spells, instants, and 'fake' casts
+   */
+  defaultHandler(
+    event: AnyEvent,
+    events: AnyEvent[],
+    eventIndex: number,
+    channelState: ChannelState,
+  ): void {
+    if (event.type === EventType.BeginCast) {
+      // we're starting a new generic hardcast
+      cancelCurrentChannel(event, channelState);
+      beginCurrentChannel(event, channelState);
```

Comment:
if the `defaultHandler` gets triggered, will this not insert a begincast for non-channeled casts?

Reply:
I'm not sure what you're asking here. The purpose of this module is that both regular hardcasts and channels are shown in the same way by the fabricated `beginchannel` and `endchannel` events. Regular hardcasts always start with a `begincast` event, and so this logic puts a `beginchannel` in the same place. This is intended behavior, and it's the same thing the module currently does on live.